airbyte-ci/connectors/pipelines/pipelines/gradle.py

```diff
@@ -121,29 +122,69 @@ async def _run(self) -> StepResult:
+                        f"./gradlew {' '.join(self.DEFAULT_GRADLE_TASK_OPTIONS)} --write-verification-metadata sha256 help --dry-run",
+                        # Store to the persistent cache.
+                        "(rsync -a --stats /root/.gradle/ /root/gradle-persistent-cache || true)",
+                        # Build the CDK and publish it to the local maven repository.
+                        # Do this last to not pollute the persistent cache.
+                        f"./gradlew {' '.join(self.DEFAULT_GRADLE_TASK_OPTIONS)} :airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded",
+                    ]
+                )
+            )
+        )
+
+        # Mount only the code needed to build the connector.
+        # This reduces the scope of the inputs to help dagger reuse container layers.
+        # The contents of '/root/.gradle' and '/root/.m2' are by design not overly sensitive to changes in the rest of the git repo.
+        gradle_container = (
+            gradle_container_base
             # TODO: remove this once we finish the project to boost source-postgres CI performance.
             .with_env_variable("CACHEBUSTER", hacks.get_cachebuster(self.context, self.logger))
-            # Mount the gradle cache volume.
-            # We deliberately don't mount it at $GRADLE_HOME, instead we load it there and store it from there using rsync.
-            # This is because the volume is accessed concurrently by all GradleTask instances.
-            # Hence, why we synchronize the writes by setting the `sharing` parameter to LOCKED.
-            .with_mounted_cache("/root/gradle-cache", self.connector_java_build_cache, sharing=CacheSharingMode.LOCKED)
-            # Mount the parts of the repo which interest us in /airbyte.
-            .with_workdir("/airbyte")
+            # Mount the connector-agnostic whitelisted files in the git repo.
             .with_mounted_directory("/airbyte", self.context.get_repo_dir(".", include=include))
+            # Mount the sources for the connector and its dependencies.
             .with_mounted_directory(str(self.context.connector.code_directory), await self.context.get_connector_dir())
-            # Disable the Ryuk container because it needs privileged docker access that does not work:
-            .with_env_variable("TESTCONTAINERS_RYUK_DISABLED", "true")
-            # Run gradle once to populate the container's local maven repository.
-            # This step is useful also to serve as a basic sanity check and to warm the gradle cache.
-            # This will download gradle itself, a bunch of poms and jars, compile the gradle plugins, configure tasks, etc.
-            .with_exec(self._get_gradle_command(":airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded"))
+            # Mount the cache volume for the transient gradle cache used for this connector only.
+            # This volume is PRIVATE meaning it exists only for the duration of the dagger pipeline.
```

Comment:
[Dagger has a `pipeline` terminology that differ from what we call a pipeline.
](https://dagger-io.readthedocs.io/en/sdk-python-v0.6.4/api.html#dagger.api.gen.Container.pipeline)
It's not clear to me if the `PRIVATE` sharing means that the cache volume will exist for the duration of a dagger session or for the duration of a sub-pipeline.
We currently have [Step subpipelines](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/bases.py#L128). 
If private cache volume exist for the duration of a sub-pipeline it will defeat your purpose of sharing the cache across steps.

Reply:
Interesting. I did find the caching to work, but now I'm starting to thing that it only worked accidentally.

Reply:
What do you think about writing an integration test that would help us actually test the caching behavior? Something using a dummy `GradleTask` subclass and  running gradle command on a dummy gradle project. It'd be helpful to validate our caching assumptions and prevent regressions...

Reply:
Kind of. I this [test](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/tests/test_tests/test_common.py#L186) I wrote some assertions about cache / cache bursting.

Reply:
Thanks!